types/plugin.d.ts

```diff
@@ -7,12 +7,11 @@ import { RawServerBase, RawServerDefault, RawRequestDefaultExpression, RawReplyD
  *
  * Fastify allows the user to extend its functionalities with plugins. A plugin can be a set of routes, a server decorator or whatever. To activate plugins, use the `fastify.register()` method.
  */
-export interface FastifyPlugin<Options extends FastifyPluginOptions = {}> {
-  (
-    instance: FastifyInstance<RawServerBase, RawRequestDefaultExpression<RawServerBase>, RawReplyDefaultExpression<RawServerBase>>,
-    opts: Options,
-    next: (err?: FastifyError) => void
-  ): void;
+export type FastifyPlugin<Options extends FastifyPluginOptions = {}, FI extends FastifyInstance = FastifyInstance> =
+{
+  (instance: FI, opts: Options, next: (err?: FastifyError) => void): void;
+} | {
```

Comment:
Lets break these apart into two separate types or interfaces. Call them `FastifyPluginSync` and `FastifyPluginAsync`. Update FastifyPlugin to be `FastifyPluginSync | FastifyPluginAsync` (pass through the options generic_, and make sure to export all three of these types. In the test file create a new test plugin that is specified by the `FastifyPluginAsync` type and pass it to register.

This _should_ fix the async bug and enable users to safely type their plugins.

Reply:
Let's call them `FastifyPluginCallback` and `FastifyPluginAsync`.  No one is sync.

Reply:
Will try this out, looks good.

Reply:
Done.